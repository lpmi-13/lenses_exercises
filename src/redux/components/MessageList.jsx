import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import classnames from "classnames";
import { List, AutoSizer } from "react-virtualized";
import { Action } from "../actions";
import ListItemDetails from "./ListItemDetails";

class MessageList extends React.Component {
  constructor(props) {
    super(props);

    this.onShowRowDetails = this.onShowRowDetails.bind(this);
    this.rowRenderer = this.rowRenderer.bind(this);

    this.state = {
      message: {}
    };
  }

  componentDidUpdate() {
    if (Object.keys(this.state.message).length === 0) {
      this.list.scrollToRow(this.props.messages.length);
    }
  }

  onShowRowDetails = d => {
    this.setState({ message: d });
  };

  rowRenderer = messages => ({
    key, // Unique key within array of rows
    index, // Index of row within collection
    isScrolling, // The List is currently being scrolled
    isVisible, // This row is visible within the List (eg it is not an overscanned row)
    style // Style object to be applied to row (to position it)
  }) => {
    let arr = [],
      keys = [];
    Object.keys(JSON.parse(messages[index].key)).forEach(function(k) {
      keys.push({ label: k, value: JSON.parse(messages[index].key)[k] });
    });
    Object.keys(JSON.parse(messages[index].value)).forEach(function(k) {
      arr.push({ label: k, value: JSON.parse(messages[index].value)[k] });
    });
    return (
      <div
        key={key}
        style={style}
        className="message-row columns ws-message-list is-multiline"
        onClick={this.onShowRowDetails.bind(this, messages[index])}
      >
        <div className="column is-2">
          <div>Index</div>
          {index}
        </div>
        {keys.map(item => (
          <MessageListItem
            className="key"
            key={item.label}
            label={item.label}
            value={item.value}
          />
        ))}
        {arr.map(item => (
          <MessageListItem
            key={item.label}
            label={item.label}
            value={item.value}
          />
        ))}
      </div>
    );
  };

  render() {
    const { messages, onCommitMessage } = this.props;
    const { onShowRowDetails } = this;
    const { message } = this.state;

    return (
      <div>
        <ListItemDetails
          message={message}
          onCommitMessage={onCommitMessage}
          onShowRowDetails={onShowRowDetails}
        />
        <nav aria-label="message panel" className="panel">
          <div className="panel-block">
            <AutoSizer className="autosizer-bulma-fix">
              {({ height, width, disableHeight = true }) => (
                <List
                  ref={list => {
                    this.list = list;
                  }}
                  width={width}
                  height={290}
                  rowCount={messages.length}
                  rowHeight={160}
                  rowRenderer={this.rowRenderer(messages)}
                />
              )}
            </AutoSizer>
          </div>
        </nav>
      </div>
    );
  }
}

class MessageListItem extends React.Component {
  render() {
    return (
      <div className="column is-2">
        <div>{this.props.label}</div>
        {this.props.value}
      </div>
    );
  }
}

MessageList.defaultProps = {};

MessageList.propTypes = {
  onCommitMessage: PropTypes.func,
  message: PropTypes.object
};

const mapStateToProps = state => ({
  message: state.session.message,
  messages: state.session.messages
});

const mapDispatchToProps = {
  ...Action
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MessageList);
