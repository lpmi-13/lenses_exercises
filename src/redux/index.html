<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="apple-mobile-web-app-capable" content="yes" />
    <meta name="mobile-web-app-capable" content="yes" />
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css">
    <title>Kafka with Lenses Demo</title>
</head>

<body>
    <div id="main" role="main"></div>
</body>
</html>
