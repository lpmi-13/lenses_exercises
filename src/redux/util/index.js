export const MESSAGE_LIMIT = 5000;
