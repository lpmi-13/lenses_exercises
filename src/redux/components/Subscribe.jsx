import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import classnames from 'classnames';
import { Actions as KafkaActions } from 'redux-lenses-streaming';
import { Action } from '../actions';
import { MESSAGE_LIMIT } from '../util'

class Subscribe extends React.Component {
  constructor(props) {
    super(props);

    this.onSqlsChange = this.onSqlsChange.bind(this);
    this.onSubscribe = this.onSubscribe.bind(this);
    this.onUnsubscribe = this.onUnsubscribe.bind(this);
    this.onClearMessages = this.onClearMessages.bind(this);

    this.state = {
      sqls: '',
    };
  }

  componentDidUpdate(props) {
    const { messages, subscriptions } = props;
    if (messages.length > MESSAGE_LIMIT) {
      subscriptions.map(topic => this.onUnsubscribe(topic))
    }
  }

  onSqlsChange(event) {
    this.setState({ sqls: event.target.value });
  }

  onSubscribe() {
    const request = {
      sqls: this.state.sqls,
    };
    this.props.subscribe(request);
  }

  onClearMessages() {
    this.props.clearMessages();
  }

  onUnsubscribe(topic) {
    const request = {
      topics: [topic],
    };
    this.props.unsubscribe(request);
  }

  render() {
    const { messages, subscriptions, connection } = this.props;
    const { sqls } = this.state;

    const btnStyle = classnames('button is-small is-info');

    const topics = subscriptions.map(subscription =>
      (
        <button
          onClick={this.onUnsubscribe.bind(this, subscription)}
          key={subscription}
          className="button is-danger is-outlined is-small is-pulled-right"
        >
          <span>{subscription}</span>
          <span className="icon is-small">
            <i className="fa fa-times" />
          </span>
        </button>)
      );

    return (
      <nav className="ws-subscribe panel">
        <div className="panel-heading">
          <div className="field has-addons">
            <p className="control is-expanded">
              <textarea
                rows="3"
                className="textarea is-small is-info"
                placeholder="SQLS"
                value={sqls}
                onChange={this.onSqlsChange}
              />
            </p>
          </div>
        </div>
        <div className="panel-block">
          <div className="control">
            <button
              style={{ marginRight: '10px' }}
              onClick={this.onSubscribe}
              className={btnStyle}
              disabled={!connection || !this.state.sqls}
            >
              Subscribe
            </button>
            <button
              onClick={this.onClearMessages}
              className="button is-small is-danger clear-messages"
              disabled={!connection}
            >
              Clear Messages
            </button>
            <button
              style={{ marginLeft: '10px' }}
              onClick={() => console.log('search')}
              className="button is-small is-success"
            >
              <span className="icon is-small">
              <i className="fas fa-check"></i>
              </span>
              <span>Search</span>
            </button>
          </div>
          <div className="control">
            {topics}
          </div>
        </div>
        <div className="panel-block">
          <div className="control number-of-messages">
            Number of messages: {messages.length > MESSAGE_LIMIT ? 0 : messages.length}
          </div>
        </div>
      </nav>
    );
  }
}

Subscribe.defaultProps = {
};

Subscribe.propTypes = {
  subscribe: PropTypes.func.isRequired,
  unsubscribe: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  subscriptions: state.lenses.subscriptions,
  connection: state.lenses.connection,
  messages: state.session.messages,
});

const mapDispatchToProps = {
  ...KafkaActions,
  ...Action,
};

export default connect(mapStateToProps, mapDispatchToProps)(Subscribe);
export { Subscribe }
