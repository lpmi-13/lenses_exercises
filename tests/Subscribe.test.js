import React from 'react';
import { Subscribe } from '../src/redux/components/Subscribe';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { MESSAGE_LIMIT } from '../src/redux/util';

describe('the Subscribe component', () => {
    
    const clearMessages = jest.fn();
    const messages = [];
    const subscribe = jest.fn();
    const unsubscribe = jest.fn();
    const subscriptions = [];

    const defaultProps = {
        clearMessages,
        messages,
        subscribe,
        subscriptions,
        unsubscribe,
    }

    it('renders completely', () => {
      const tree = renderer.create(<Subscribe {...defaultProps} />).toJSON();
      expect(tree).toMatchSnapshot();        
    });

    it('clears all messages when Clear Messages is clicked', () => {
        const messages = ['this is a message', 'this is also a message'];
        const propsWithMessages = {...defaultProps, messages};
        const component = shallow(<Subscribe { ...propsWithMessages } />);
        expect(component.find('.number-of-messages').text()).toBe('Number of messages: 2');
        component.find('.clear-messages').simulate('click');
        expect(clearMessages).toHaveBeenCalled(); 
    });

    it('only shows messages up to the constant value set in utils', () => {
        const lotsOfMessages = Array(3000).fill('message');
        const propsWithLotsOfMessages = {...defaultProps, messages: lotsOfMessages};
        const component = shallow(<Subscribe {...propsWithLotsOfMessages } />);
        expect(component.find('.number-of-messages').text()).toBe(`Number of messages: ${lotsOfMessages.length}`);

        const tooManyMessages = Array(MESSAGE_LIMIT + 1).fill('message');
        component.setProps({ messages: tooManyMessages });
        expect(component.find('.number-of-messages').text()).toBe('Number of messages: 0');
    })
})