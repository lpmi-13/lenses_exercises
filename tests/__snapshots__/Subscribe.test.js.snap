// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Subscribe component renders completely 1`] = `
<nav
  className="ws-subscribe panel"
>
  <div
    className="panel-heading"
  >
    <div
      className="field has-addons"
    >
      <p
        className="control is-expanded"
      >
        <textarea
          className="textarea is-small is-info"
          onChange={[Function]}
          placeholder="SQLS"
          rows="3"
          value=""
        />
      </p>
    </div>
  </div>
  <div
    className="panel-block"
  >
    <div
      className="control"
    >
      <button
        className="button is-small is-info"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "marginRight": "10px",
          }
        }
      >
        Subscribe
      </button>
      <button
        className="button is-small is-danger clear-messages"
        disabled={true}
        onClick={[Function]}
      >
        Clear Messages
      </button>
      <button
        className="button is-small is-success"
        onClick={[Function]}
        style={
          Object {
            "marginLeft": "10px",
          }
        }
      >
        <span
          className="icon is-small"
        >
          <i
            className="fas fa-check"
          />
        </span>
        <span>
          Search
        </span>
      </button>
    </div>
    <div
      className="control"
    />
  </div>
  <div
    className="panel-block"
  >
    <div
      className="control number-of-messages"
    >
      Number of messages: 
      0
    </div>
  </div>
</nav>
`;
